### PR TITLE
[WIP] Allow all tfr_functions to operate on Raw object

### PR DESCRIPTION
**Pull request by adam2392**
#### Reference issue
Addresses tfr discussion in: #8773 

#### What does this implement/fix?
1. Allows `tfr` functions to operate on ``Raw`` object that returns and `AverageTFR` object.
2. ~~Adds an API to `AverageTFR` for plotting a spectrogram and averaging frequencies between two ranges~~

#### Additional information
~~API is draft just to show what it would be like. I have not implemented 2. yet, and can take that out if you think it is not necessary/useful.~~

Related discussion on discourse: https://mne.discourse.group/t/is-there-a-way-to-obtain-the-time-points-from-time-frequency-stft/2682/3

### CODE REVIEW

**Comment by adam2392** [comment]
@agramfort this is something along what I would propose for what we disucced in #8773

**Comment by adam2392** [comment]
@agramfort just wanted to check in with you to see if this is going in the right direction? The main point is to allow tfr functions to operate on `Raw` objects with a disclaimer in the docstring about size/runtime.

Then a few questions regarding unimplemented functions:
- `average_freqs`: the intention is to provide a way to average within the frequency range too for things like computing "gamma band power", or something like that.
- `plot_spectrogram`: the intention here is to provide a plotting mechanism for plotting channels X time with frequency power (possibly in dB) as the color

^ Are these two functionalities desirable in the `_BaseTFR` class?

**Comment by agramfort** [comment]
>
> @agramfort <https://github.com/agramfort> just wanted to check in with
> you to see if this is going in the right direction? The main point is to
> allow tfr functions to operate on Raw objects with a disclaimer in the
> docstring about size/runtime.
>
> Then a few questions regarding unimplemented functions:
>
>    - average_freqs: the intention is to provide a way to average within
>    the frequency range too for things like computing "gamma band power", or
>    something like that.
>
>
for this usecase I would raw.filter(...).apply_hilbert(envelope=False)
it's much faster than an STFT and averaging. It's also much easier to
explain.


>
>    - plot_spectrogram: the intention here is to provide a plotting
>    mechanism for plotting channels X time with frequency power (possibly in
>    dB) as the color
>
> ^ Are these two functionalities desirable in the _BaseTFR class?
>

how would it look like? you would display minutes of data? would you have a
way to scroll?

**Comment by adam2392** [comment]
>The main point is to allow tfr functions to operate on Raw objects with a disclaimer in the docstring about size/runtime. Then a few questions regarding unimplemented functions: - average_freqs: the intention is to provide a way to average within the frequency range too for things like computing "gamma band power", or something like that.
> for this usecase I would raw.filter(...).apply_hilbert(envelope=False) it's much faster than an STFT and averaging. It's also much easier to explain.

I might not be an expert in time-freq signal processing, but aren't there some advantages to using FFT like techniques and then averaging (https://mne.tools/stable/auto_examples/time_frequency/plot_time_frequency_simulated.html), or is Hilbert Transform envelope considered the best for showing things like frequency bands? 

I might be lacking some basic working knowledge here... Lmk if instead you would prefer I post a more general question on the Discourse.

> - plot_spectrogram: the intention here is to provide a plotting mechanism for plotting channels X time with frequency power (possibly in dB) as the color ^ Are these two functionalities desirable in the _BaseTFR class?
> how would it look like? you would display minutes of data? would you have a way to scroll?
 
It would just be a static heatmap

**Comment by agramfort** [comment]
>
> I might not be an expert in time-freq signal processing, but aren't there
> some advantages to using FFT like techniques and then averaging (
> https://mne.tools/stable/auto_examples/time_frequency/plot_time_frequency_simulated.html),
> or is Hilbert Transform envelope considered the best for showing things
> like frequency bands?
>
do your own experiments if you have doubts.

> I might be lacking some basic working knowledge here... Lmk if instead you
> would prefer I post a more general question on the Discourse.
>
>
>    - plot_spectrogram: the intention here is to provide a plotting
>    mechanism for plotting channels X time with frequency power (possibly in
>    dB) as the color ^ Are these two functionalities desirable in the _BaseTFR
>    class?
>    how would it look like? you would display minutes of data? would you
>    have a way to scroll?
>
> It would just be a static heatmap
>

honestly I would encourage you to develop this tool for your own needs and
once doing
a few iterations on your side  see what's best for MNE-Python.

**Comment by adam2392** [comment]
Okay sounds good! Thanks @agramfort for answering questions. I've updated the PR now to just reflect the title of allowing TFR funcs to operate on Raw objects. Lmk if this is good?

File: mne/time_frequency/tfr.py [hunk 1]
```diff
@@ -872,6 +882,12 @@ def tfr_multitaper(inst, freqs, n_cycles, time_bandwidth=4.0,
     Notes
     -----
     .. versionadded:: 0.9.0
+
+    If the function is run on a ``Raw`` object, then it one should be
```
Comment (agramfort) [inline mne/time_frequency/tfr.py:886]:
```suggestion
    If the function is run on a ``Raw`` object, then it should be
```

File: mne/time_frequency/tfr.py [hunk 2]
```diff
@@ -728,6 +730,14 @@ def tfr_morlet(inst, freqs, n_cycles, use_fft=False, return_itc=True, decim=1,
     mne.time_frequency.tfr_array_multitaper
     mne.time_frequency.tfr_stockwell
     mne.time_frequency.tfr_array_stockwell
+
+    Notes
+    -----
+    If the function is run on a ``Raw`` object, then it one should be
```
Comment (agramfort) [inline mne/time_frequency/tfr.py:736]:
```suggestion
    If the function is run on a ``Raw`` object, then it should be
```

**Review by agramfort — COMMENTED** [review]
you miss a what's new entry and I would strongly recommend to update a tutorial on time frequency to document this option

File: mne/time_frequency/tfr.py [hunk 3]
```diff
@@ -2173,13 +2189,17 @@ def combine_tfr(all_tfr, weights='nave'):
...
     from ..evoked import Evoked
-    if not isinstance(inst, (BaseEpochs, Evoked)):
-        raise TypeError('inst must be Epochs or Evoked')
+    from ..io import BaseRaw
+    if not isinstance(inst, (BaseEpochs, Evoked, BaseRaw)):
+        raise TypeError('inst must be Epochs or Evoked or Raw')
     if isinstance(inst, BaseEpochs):
         data = inst.get_data()
+    elif isinstance(inst, BaseRaw):
+        # make data into 3D to play nice with epochs x ch x time
```
Comment (agramfort) [inline mne/time_frequency/tfr.py:2201]: should you also do this:

https://github.com/mne-tools/mne-python/blob/8bc55d7435ce047c37465f2b5c173517491d7002/mne/time_frequency/tfr.py#L2204

I am thinking that maybe you can use the same code path for Evoked and Raw? They basically both contain 2D data arrays

Comment (adam2392) [reply mne/time_frequency/tfr.py]: Like what's done here: c14d080?

**Comment by adam2392** [comment]
Added an extension to an existing tutorial. Lmk if you think this is good, or if I should add more?

**Comment by agramfort** [comment]
@adam2392 can you share some rendered doc page when you have one on CI? thx

**Pull request closed without merging**